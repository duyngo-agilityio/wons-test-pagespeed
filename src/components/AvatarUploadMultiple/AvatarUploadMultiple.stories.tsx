import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

// Components
import AvatarUploadMultiple from '.';

const meta: Meta<typeof AvatarUploadMultiple> = {
  title: 'Components/AvatarUploadMultiple',
  component: AvatarUploadMultiple,
  parameters: {
    layout: 'centered',
    controls: { expanded: true },
  },
};

export default meta;
type Story = StoryObj<typeof AvatarUploadMultiple>;

const mockPreviewFile = [
  'https://img-cdn.pixlr.com/image-generator/history/65bb506dcb310754719cf81f/ede935de-1138-4f66-8ed7-44bd16efc709/medium.webp',
  'https://images.unsplash.com/photo-1505740420928-5e560c06d30e?fm=jpg&q=60&w=3000&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8Nnx8cHJvZHVjdHxlbnwwfHwwfHx8MA%3D%3D',
];

export const Default: Story = {
  args: {
    previewFiles: mockPreviewFile,
    onFileChange: fn(),
  },
};
