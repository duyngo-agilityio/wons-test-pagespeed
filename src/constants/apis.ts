export const API_BASE_URL = process.env.NEXT_API_BASE_URL || '';

export const API_PATH = {
  SIGN_UP: '/auth/local/register',
  USERS: '/users',
};
