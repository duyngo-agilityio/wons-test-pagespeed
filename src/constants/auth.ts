// Types
import { TRole } from '@/types';

export const AUTH_METHODS = {
  CREDENTIALS: 'credentials',
};

export const ERROR_TYPES = {
  CREDENTIALS_SIGN_IN: 'CallbackRouteError',
};

export const ROLES: TRole[] = [
  {
    id: 3,
    name: 'Admin',
  },
  {
    id: 4,
    name: 'User',
  },
];
