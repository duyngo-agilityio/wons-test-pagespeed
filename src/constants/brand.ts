export const BRANDS = [
  { key: 'apple', label: 'Apple' },
  { key: 'samsung', label: 'Samsung' },
  { key: 'huawei', label: 'Huawei' },
  { key: 'xioami', label: 'Xioami' },
  { key: 'oppo', label: 'Oppo' },
  { key: 'google', label: 'Google' },
];
