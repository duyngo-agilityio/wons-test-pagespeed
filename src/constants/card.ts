export enum LevelCard {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
}
