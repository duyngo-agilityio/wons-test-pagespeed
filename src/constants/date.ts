export const DEFAULT_RANGE_VALUE_PICKER = 7;
