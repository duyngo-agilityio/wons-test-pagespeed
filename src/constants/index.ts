export * from './apis';
export * from './auth';
export * from './date';
export * from './fonts';
export * from './images';
export * from './messages';
export * from './params';
export * from './regex';
export * from './routes';
export * from './sidebar';
export * from './themes';
export * from './typography';
export * from './messages';
export * from './regex';
export * from './apis';
export * from './auth';
export * from './invoices';
export * from './select';
export * from './pagination';
