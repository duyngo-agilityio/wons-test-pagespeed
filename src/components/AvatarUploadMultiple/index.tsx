'use client';

import { ChangeEvent, useCallback, useState } from 'react';
import { IoCamera, IoClose } from 'react-icons/io5';
import { Input, Button, Text, ImageFallback } from '@/components';

// constants
import { ERROR_MESSAGES, MAX_SIZE, REGEX } from '@/constants';

export type TAvatarUploadMultipleProps = {
  onFileChange: (files: File[]) => void;
  value?: string[];
  error?: string;
};

const AvatarUploadMultiple = ({
  value = [],
  error = '',
  onFileChange,
}: TAvatarUploadMultipleProps) => {
  const [previewURLs, setPreviewURLs] = useState<string[]>(value);
  const [errorMessage, setErrorMessage] = useState<string>(error);
  const [_, setFilesList] = useState<File[]>([]);

  const handleChangeFiles = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      const files = e.target.files;
      if (!files) return;

      // Reset error message
      setErrorMessage('');

      if (files.length > 2) {
        setErrorMessage(ERROR_MESSAGES.MAX_IMAGE);
        return;
      }

      // Array to hold valid files and preview URLs
      const validFiles: File[] = [];
      const newPreviewURLs: string[] = [];

      // Validate files
      for (let i = 0; i < files.length; i++) {
        const file = files[i];

        // Validate file name and size
        if (!REGEX.IMG.test(file.name)) {
          setErrorMessage(ERROR_MESSAGES.UPLOAD_IMAGE);
          return;
        }
        if (file.size > MAX_SIZE) {
          setErrorMessage(ERROR_MESSAGES.UPLOAD_IMAGE_SIZE);
          return;
        }

        // Add valid file
        validFiles.push(file);
        const previewImage = URL.createObjectURL(file);
        newPreviewURLs.push(previewImage);
      }

      // Use the callback version of setState to ensure we have the latest value
      setFilesList((prevFiles) => {
        const updatedFiles = [...prevFiles, ...validFiles];

        // Pass the updated files list to the parent component
        onFileChange(updatedFiles);

        return updatedFiles; // Update the filesList state
      });

      // Update preview URLs
      setPreviewURLs((prevURLs) => [...prevURLs, ...newPreviewURLs]);
    },
    [onFileChange], // Remove filesList from dependency array
  );

  const handleRemoveImage = (index: number) => {
    // Remove the image from previewURLs
    const updatedPreviews = previewURLs.filter((_, i) => i !== index);
    setPreviewURLs(updatedPreviews);

    // Also remove the corresponding file from filesList
    setFilesList((prevFiles) => {
      const updatedFiles = prevFiles.filter((_, i) => i !== index);
      // Pass the updated files list to the parent component
      onFileChange(updatedFiles);
      return updatedFiles;
    });
  };

  const isUploadDisabled = previewURLs.length >= 2;

  return (
    <div className="flex flex-col items-center mt-5">
      <label
        htmlFor="file"
        className={`flex items-center justify-center w-full h-20 border-2 border-dashed border-gray-300 rounded-lg cursor-pointer transition-all p-11 ${
          isUploadDisabled
            ? 'cursor-not-allowed opacity-50'
            : 'hover:border-blue-500'
        }`}
      >
        <IoCamera size={24} className="text-blue-800/70" />
        <Text className="ml-2" text="Upload Image" />
      </label>

      <Input
        aria-label="Upload Avatar"
        type="file"
        id="file"
        className="hidden"
        accept="image/*"
        multiple
        onChange={handleChangeFiles}
        disabled={isUploadDisabled}
      />

      {errorMessage && (
        <p className="text-red-500 text-sm mt-2">{errorMessage}</p>
      )}

      <div className="grid grid-cols-2 gap-4 mt-4">
        {previewURLs.map((url, index) => (
          <div key={index} className="relative w-24 h-24">
            <ImageFallback
              src={url}
              alt={`Uploaded image ${index + 1}`}
              width={96}
              height={96}
              className="rounded-md object-cover w-full h-full"
            />
            <Button
              className="absolute top-0 right-0 p-1 !border-none rounded-none !bg-transparent dark:!bg-transparent !text-pink-500"
              onClick={() => handleRemoveImage(index)}
            >
              <IoClose size={16} />
            </Button>
          </div>
        ))}
      </div>
    </div>
  );
};

export default AvatarUploadMultiple;
