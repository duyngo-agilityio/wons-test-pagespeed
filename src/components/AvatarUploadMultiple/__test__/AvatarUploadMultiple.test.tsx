import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import AvatarUploadMultiple, { TAvatarUploadMultipleProps } from '../index';

// Constants
import { MESSAGES } from '@/constants';

describe('AvatarUploadMultiple Component', () => {
  const mockOnFileChange = jest.fn();

  const defaultProps: TAvatarUploadMultipleProps = {
    previewFiles: [],
    onFileChange: mockOnFileChange,
  };

  const setup = (props = defaultProps) =>
    render(<AvatarUploadMultiple {...props} />);

  beforeEach(() => {
    global.URL.createObjectURL = jest.fn(
      (file: File) => `/preview-${file.name}`,
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the upload button and no preview images when initialized', () => {
    setup();
    expect(screen.getByText('Upload Image')).toBeInTheDocument();
    expect(screen.queryByRole('img')).not.toBeInTheDocument();
  });

  it('uploads multiple files and displays previews', () => {
    setup();

    const input = screen.getByLabelText('Upload Avatar') as HTMLInputElement;

    const file1 = new File(['content'], 'file1.jpg', { type: 'image/jpeg' });
    const file2 = new File(['content'], 'file2.jpg', { type: 'image/jpeg' });

    fireEvent.change(input, { target: { files: [file1, file2] } });

    expect(screen.getAllByRole('img').length).toBe(2);
    expect(screen.getByAltText('Uploaded image 1')).toHaveAttribute(
      'src',
      'http://localhost/_next/image?url=%2Fpreview-file1.jpg&w=256&q=75',
    );
    expect(screen.getByAltText('Uploaded image 2')).toHaveAttribute(
      'src',
      'http://localhost/_next/image?url=%2Fpreview-file2.jpg&w=256&q=75',
    );
    expect(mockOnFileChange).toHaveBeenCalledWith(
      ['/preview-file1.jpg', '/preview-file2.jpg'],
      [file1, file2],
    );
  });

  it('disables the upload button when two files are already uploaded', () => {
    setup({
      ...defaultProps,
      previewFiles: ['/preview-file1.jpg', '/preview-file2.jpg'],
    });

    const input = screen.getByLabelText('Upload Avatar') as HTMLInputElement;
    expect(input).toBeDisabled();
  });

  it('shows error message when attempting to upload more than two files', () => {
    setup({
      ...defaultProps,
      previewFiles: ['/preview-file1.jpg'],
    });

    const input = screen.getByLabelText('Upload Avatar') as HTMLInputElement;

    const file1 = new File(['content'], 'file1.jpg', { type: 'image/jpeg' });
    const file2 = new File(['content'], 'file2.jpg', { type: 'image/jpeg' });
    const file3 = new File(['content'], 'file3.jpg', { type: 'image/jpeg' });

    fireEvent.change(input, { target: { files: [file1, file2, file3] } });

    expect(
      screen.getByText(MESSAGES.ERROR.UPLOAD_IMAGE_SIZE),
    ).toBeInTheDocument();
    expect(mockOnFileChange).not.toHaveBeenCalled();
  });

  it('removes a file preview when delete button is clicked', () => {
    setup({
      ...defaultProps,
      previewFiles: ['/preview-file1.jpg', '/preview-file2.jpg'],
    });

    const deleteButtons = screen.getAllByRole('button', { hidden: true });
    fireEvent.click(deleteButtons[0]);

    expect(mockOnFileChange).toHaveBeenCalledWith(
      ['/preview-file2.jpg'],
      expect.any(Array),
    );
  });

  it('clears input value on click to allow re-uploading the same file', () => {
    setup();

    const input = screen.getByLabelText('Upload Avatar') as HTMLInputElement;

    fireEvent.change(input, {
      target: {
        files: [new File(['content'], 'file.jpg', { type: 'image/jpeg' })],
      },
    });
    expect(input.files?.length).toBe(1);

    fireEvent.click(input);
    expect(input.value).toBe('');
  });

  it('matches the snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
