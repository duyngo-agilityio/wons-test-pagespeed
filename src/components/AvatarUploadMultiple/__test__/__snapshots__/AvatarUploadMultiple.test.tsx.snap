// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarUploadMultiple Component matches the snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center mt-5"
  >
    <label
      class="flex items-center justify-center w-full h-20 border-2 border-dashed border-gray-300 rounded-lg cursor-pointer transition-all p-11 hover:border-blue-500"
      for="file"
    >
      <svg
        class="text-blue-800/70"
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="256"
          cy="272"
          r="64"
        />
        <path
          d="M432 144h-59c-3 0-6.72-1.94-9.62-5l-25.94-40.94a15.52 15.52 0 0 0-1.37-1.85C327.11 85.76 315 80 302 80h-92c-13 0-25.11 5.76-34.07 16.21a15.52 15.52 0 0 0-1.37 1.85l-25.94 41c-2.22 2.42-5.34 5-8.62 5v-8a16 16 0 0 0-16-16h-24a16 16 0 0 0-16 16v8h-4a48.05 48.05 0 0 0-48 48V384a48.05 48.05 0 0 0 48 48h352a48.05 48.05 0 0 0 48-48V192a48.05 48.05 0 0 0-48-48zM256 368a96 96 0 1 1 96-96 96.11 96.11 0 0 1-96 96z"
        />
      </svg>
      <p
        class="font-dm-sans text-blue-800 dark:text-white text-md ml-2"
      >
        Upload Image
      </p>
    </label>
    <div
      class="group flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[29px] hidden"
      data-filled="true"
      data-filled-within="true"
      data-slot="base"
    >
      <div
        class="h-full flex flex-col"
        data-slot="main-wrapper"
      >
        <div
          class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm gap-3 min-h-10 transition-background motion-reduce:transition-none !duration-150 data-[hover=true]:bg-primary-100 text-primary placeholder:text-primary outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-offset-background bg-gray-50 dark:bg-gray-600 hover:data-[hover=true]:bg-gray-200/50 dark:hover:bg-gray-900 focus-within:bg-gray-50 dark:focus-within:bg-gray-600 group-data-[focus=true]:bg-gray-50 dark:group-data-[focus=true]:bg-gray-600 h-12.5 p-3.75 rounded-[10px] group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-blue-500"
          data-slot="input-wrapper"
          style="cursor: text;"
        >
          <div
            class="inline-flex w-full items-center h-full box-border"
            data-slot="inner-wrapper"
          >
            <input
              accept="image/*"
              aria-label="Upload Avatar"
              class="w-full font-normal !outline-none focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-blue-800/70 dark:text-white/70 placeholder:text-blue-800/40 dark:placeholder:text-white/40 bg-transparent text-md"
              data-slot="input"
              id="file"
              multiple=""
              placeholder=" "
              type="file"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="grid grid-cols-2 gap-4 mt-4"
    />
  </div>
</div>
`;
